### PR TITLE
Updated auth module as per the interop requirements of the with AD server

### PR DESCRIPTION
**Pull request by maheshkelkar**
* Authentication is performed in 4 step process
  * Create a context (A) using bind credentials
  * Search userDN with the context (A) and username
  * Rebind or create a new context (B) using userDN and password
  * Search groups with context (A) and userDN
* Added informative javadocs
* Updated the configuration to fall in line with the requirements
* Moved the Test LDAP server to common test code
* Updated the LdapAuthenticator Spock tests to use TestLdapServer
* Save DeployDBConfigurtion object in the DeployDB app
* Added manualPromo.yml and advServ.yml for testing manual promotion

References #164

### CODE REVIEW

**Comment by maheshkelkar** [comment]
@GiriDandu please review

File: example/modelConfig/promotions/manualPromo.yml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+type:  deploydb.models.promotion.ManualLDAPPromotionImpl
+description: "Manual LDAP Promotion"
+attributes:
+    allowedGroup: GRP-AWS-Dev-Eng
```
Comment (maheshkelkar) [inline example/modelConfig/promotions/manualPromo.yml]: Change this to generic group name

**Comment by maheshkelkar** [comment]
@rtyler @GiriDandu Please take a look now. Travis is happy :)

File: dropwizard-integtest/build.gradle [hunk 2]
```diff
@@ -30,6 +30,9 @@ dependencies {
     compile ('io.dropwizard.modules:dropwizard-flyway:0.7.0-1')
 
     compile ('joda-time:joda-time:2.6+')
+
+    /* Include unboundid sdk for in-memory ldap test server */
+    compile 'com.unboundid:unboundid-ldapsdk:2.3.8'
```
Comment (rtyler) [inline dropwizard-integtest/build.gradle]: This is putting a DeployDB specific dependency into the dropwizard-integtest which I want to tear out at some poiint. This is okay temporarily but I expect @GiriDandu to remove it in #167

File: src/test/groovy/deploydb/auth/LdapAuthenticatorSpec.groovy [hunk 3]
```diff
@@ -102,116 +135,113 @@ class LdapAuthenticatorWithArgsSpec extends Specification {
...
         given:
-        ldapConfiguration.groupNameAttribute = "pizza"
         LdapAuthenticator ldapAuthenticator = new LdapAuthenticator(ldapConfiguration)
-
-        /**
-         * Mock the interfaces to return following attributes
-         */
-        InitialDirContext context = mockGetGroupMembershipInterface("pizza", "Fandango")
+        InitialDirContext context = ldapAuthenticator.bindContext()
+        final String filter = String.format("(&(%s=%s)(objectClass=%s))",
```
Comment (rtyler) [inline src/test/groovy/deploydb/auth/LdapAuthenticatorSpec.groovy]: This is the same exact logic from LdapAuthenticator L242, this is a *STRONG* indication that this logic should be refactored into a method that you can call inside of this test, and inside of that authenticateUser method

File: src/test/groovy/deploydb/auth/LdapAuthenticatorSpec.groovy [hunk 4]
```diff
@@ -102,116 +135,113 @@ class LdapAuthenticatorWithArgsSpec extends Specification {
...
-        ldapConfiguration.groupNameAttribute = "pizza"
+        ldapConfiguration.distinguishedNamePrefix = "invalidDN"
         LdapAuthenticator ldapAuthenticator = new LdapAuthenticator(ldapConfiguration)
-
-        /**
-         * Mock the interfaces to return following attributes
-         */
-        InitialDirContext context = mockGetGroupMembershipInterface("BadPizza", "Fandango")
+        InitialDirContext context = ldapAuthenticator.bindContext()
+        final String filter = String.format("(&(%s=%s)(objectClass=%s))",
```
Comment (rtyler) [inline src/test/groovy/deploydb/auth/LdapAuthenticatorSpec.groovy]: Ditto to above about refactoring

File: src/test/groovy/deploydb/auth/LdapAuthenticatorSpec.groovy (hunk 3 above)
Comment (maheshkelkar) [inline src/test/groovy/deploydb/auth/LdapAuthenticatorSpec.groovy]: ok. I moved it to a function :)

File: src/test/groovy/deploydb/auth/LdapAuthenticatorSpec.groovy (hunk 4 above)
Comment (maheshkelkar) [inline src/test/groovy/deploydb/auth/LdapAuthenticatorSpec.groovy]: done

**Pull request merged**